lib/ansible/modules/cloud/amazon/ec2_vol.py

```diff
@@ -221,150 +224,160 @@
             if name:
                 msg += (" and name: %s" % name)
             module.fail_json(msg=msg)
         else:
             return None
 
     if len(vols) > 1:
-        module.fail_json(msg="Found more than one volume in zone (if specified) with name: %s" % name)
-    return vols[0]
+        module.fail_json(
+            msg="Found more than one volume in zone (if specified) with name: %s" % name,
+            found=[v['VolumeId'] for v in vols]
+        )
+    vol = camel_dict_to_snake_dict(vols[0])
+    return vol
 
 
-def get_volumes(module, ec2):
+def get_volumes(module, ec2_conn):
 
     instance = module.params.get('instance')
 
-    try:
-        if not instance:
-            vols = ec2.get_all_volumes()
-        else:
-            vols = ec2.get_all_volumes(filters={'attachment.instance-id': instance})
-    except boto.exception.BotoServerError as e:
-        module.fail_json(msg="%s: %s" % (e.error_code, e.error_message))
-    return vols
-
+    find_params = dict()
+    if instance:
+        find_params['Filters'] = ansible_dict_to_boto3_filter_list({'attachment.instance-id': instance})
 
-def delete_volume(module, ec2):
-    volume_id = module.params['id']
+    vols = []
     try:
-        ec2.delete_volume(volume_id)
-        module.exit_json(changed=True)
-    except boto.exception.EC2ResponseError as ec2_error:
-        if ec2_error.code == 'InvalidVolume.NotFound':
-            module.exit_json(changed=False)
-        module.fail_json(msg=ec2_error.message)
-
-
-def boto_supports_volume_encryption():
-    """
-    Check if Boto library supports encryption of EBS volumes (added in 2.29.0)
-
-    Returns:
-        True if boto library has the named param as an argument on the request_spot_instances method, else False
-    """
-    return hasattr(boto, 'Version') and LooseVersion(boto.Version) >= LooseVersion('2.29.0')
-
+        vols_response = ec2_conn.describe_volumes(**find_params)
+        vols = [camel_dict_to_snake_dict(vol) for vol in vols_response.get('Volumes', [])]
+    except (botocore.exceptions.ClientError, botocore.exceptions.BotoCoreError) as e:
+        module.fail_json_aws(e, msg='Error while getting EBS volumes')
+    return vols
 
-def boto_supports_kms_key_id():
-    """
-    Check if Boto library supports kms_key_ids (added in 2.39.0)
 
-    Returns:
-        True if version is equal to or higher then the version needed, else False
-    """
-    return hasattr(boto, 'Version') and LooseVersion(boto.Version) >= LooseVersion('2.39.0')
+def delete_volume(module, ec2_conn, volume_id=None):
+    changed = False
+    if volume_id:
+        try:
+            ec2_conn.delete_volume(VolumeId=volume_id)
+            changed = True
+        except (botocore.exceptions.ClientError, botocore.exceptions.BotoCoreError) as e:
+            if e.response['Error']['Code'] != 'InvalidVolume.NotFound':
+                module.fail_json_aws(e, msg='Error while deleting volume')
```

Comment:
It seems like this could happen due to a race condition. If the volume was in the process of being deleted but still existed at the start of the module run so re-deleting fails, shouldn't this just pass and allow changed = False? Or are there other circumstances that cause this?